agents_infra/agents_infra/agents/dns/dns.py

```diff
@@ -1,33 +1,40 @@
-import logging
+import abc
 import subprocess
 
-from ...utils.helpers import is_valid_ip, restart_service
+from ...utils.configtools import Config
+from ...utils.helpers import is_valid_ip
 from ...utils.logtools import maybe_log_message
+from ...utils.sysinfo import is_port_open
 from ..base import ServerAgent
 
 
 class DNSAgent(ServerAgent):
 
+    __metaclass__ = abc.ABCMeta
+
     def __init__(
         self,
-        server_name='dns',
-        port=53,
-        processes=None,
-        critical_processes=None,
-        interface=None,
         protocol='udp',
-        whitelist_commands=None,
         command_queue_size=0,
+        config=None
     ):
+        # If not config - set default
+        if config is None:
+            config = Config(name='dns', protocol=protocol)
+        elif isinstance(config, dict):
+            config.setdefault('name', 'dns')
+            config.setdefault('protocol', protocol)
+            config = Config.from_dict(config)
+        elif isinstance(config, Config):
+            if not config.name:
+                config.name = 'dns'
+            if not getattr(config, 'protocol', None):
+                config.protocol = protocol
```

Comment:
While there is nothing wrong with doing extra attribute checks and forcing specific values in each case, I wouldn't do it as it's not very 'Pythonic'. It'd better allow the user to pass whatever they want here and later deal with consequences and debug based on logs. In fewer words, what I suggest is the following:

```python
if config is None:
    config = Config(name='dns', protocol=protocol)
elif isinstance(config, dict):
    config = Config.from_dict(config)
```

And that's it.